File: 496. Next Greater Element I/Python/Solution.py
class Solution(object):
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        result = []
        if not findNums:
            return result
        num_max = max(nums)
        map = [-1] * (num_max + 1)
        for index, num in enumerate(nums):
            map[num] = index
        for find_num in findNums:
            if find_num >= num_max:
                result.append(-1)
            else:
                current_index = map[find_num] + 1
                while current_index < len(nums):
                    if nums[current_index] > find_num:
                        result.append(nums[current_index])
                        break
                    current_index += 1
                if current_index == len(nums):
                    result.append(-1)
        return result

File: 496. Next Greater Element I/Java/Solution.java
public class Solution {
    public int[] nextGreaterElement(int[] findNums, int[] nums) {
        if (findNums.length == 0) {
            return new int[] {};
        }
        int numMax = -1;
        for (int num : nums) {
            if (num > numMax) {
                numMax = num;
            }
        }
        int[] map = new int[numMax + 1];
        for (int i = 0; i < nums.length; i++) {
            map[nums[i]] = i;
        }
        int[] result = new int[findNums.length];
        for (int i = 0; i < findNums.length; i++) {
            int currentNum = findNums[i];
            if (currentNum > numMax) {
                result[i] = -1;
            } else {
                int currentIndex = map[currentNum] + 1;
                while (currentIndex < nums.length) {
                    if (nums[currentIndex] > currentNum) {
                        result[i] = nums[currentIndex];
                        break;
                    }
                    currentIndex++;
                }
                if (currentIndex == nums.length) {
                    result[i] = -1;
                }
            }
        }
        return result;
    }
}